[fix] Fixed redundant node labels and overlays #454

Dhruv-ub opened the pull request:
To improve the interaction between map labels and hover overlays to avoid visual clutter and interaction conflicts, especially at higher zoom levels.
The hover overlay is confirmed to be the default ECharts tooltip. Its behavior is made more predictable, while label rendering is adjusted so labels no longer interfere with hover interactions.

Fixes #454

## Checklist

- [x] I have read the [OpenWISP Contributing Guidelines](http://openwisp.io/docs/developer/contributing.html).
- [x] I have manually tested the changes proposed in this pull request.
- [x] I have written new test cases for new code and/or updated existing tests for changes to existing code.
- [x] I have updated the documentation.

## Reference to Existing Issue

Closes #454.

## Description of Changes

- Clarified and stabilized tooltip behavior (default ECharts tooltip, confined to viewport, no hide delay).
- Introduced showMapLabelsAtZoom to explicitly control when geographic map labels are displayed.
- Preserved backward compatibility with showLabelsAtZoomLevel.
- Made both graph and map labels non-interactive (silent) to prevent overlap and hover conflicts.
- Ensured labels remain a passive, non-invasive layer while tooltips remain the primary interactive overlay.
- Aligned configuration naming between graph and map label zoom behavior.

## Screenshot

![IMG20251216200740](https://github.com/user-attachments/assets/b2c0acb5-8126-48e4-a4e3-6b8be17abb2a)

nemesifier reviewed (changes requested):
The patch contains changes to the formatting which are polluting the actual changes. If formatting needs to be updated let's open a PR to master with only formatting changes.

nemesifier reviewed (changes requested):
I need to insist, please only send the bugfix change, do not send any other change, if QA checks fail, we'll look into it. The PR in the current state contains a lot of other changes unrelated to the bugfix which is time consuming for maintaienrs to review.

Dhruv-ub commented:
Done please have a look on why ci build failing.

nemesifier commented:
@coderabbitai review

stktyagi commented:
> @stktyagi Thanks for suggestion but I already ran that command, still I am having issue.

Can you show me what you see after running `openwisp-qa-format`? 
Let's talk about this in the dev chat to keep this PR with comments only related to the main issue.

stktyagi commented:
> Hey, @Dhruv-ub try running `openwisp-qa-format` to clear any formatting issues. You can check if it worked by running the run-qa-checks file, `./run-qa-checks` next. If it shows no error you are good to go.

This worked, you might want to recheck your setup environment.

nemesifier reviewed (changes requested):
I see progress but I also see more quirks:

Placing the mouse on the small label causes it start blinking:

![Image](https://github.com/user-attachments/assets/d6faaf1f-262f-471e-a422-15c495038907)

When the mouse leaves, there's still overlap:

![Image](https://github.com/user-attachments/assets/08d41e56-f79c-4ce2-9a4c-1d0ee2e46272)

Please perform rigorous manual testing of your changes to ensure they're not adding quirks before asking for further reviews.

If we get to something that works, it would be nice to add at least 1 browser test for this.

Can you reply to the questions listed in the issue descirption and how you intend to solve each?

> I don't have a clear directive now, but I think the whole situation is a mess, here's the problems I see, I make suggestions at each point but I am open for better ideas:
> 
> - I am not sure where the hover overlay is coming from, whether is enabled by default or not
> - It seems to me there's not an explicit ways to disable labels on the geographic map (they appears automatically after a certain level of zoom)
> - If both are enabled, the overlap covers the label and they're quite redundant, ideally the labels should disappear if there's another overlay showing up on hover and by default only one of the two should be enabled (probably the label as it's not invasive)
> - The current config options for controlling this behavior are: `showLabelsAtZoomLevel` for geo map and `showGraphLabelsAtZoom`, which is not consistent, I think we should rename `showLabelsAtZoomLevel` to `showMapLabelsAtZoom` (in the code we can continue supporting `showLabelsAtZoomLevel` for backward compatibility) and allow this to be `false` to disable lables altogether

Dhruv-ub commented:
> I see progress but I also see more quirks:
> 
> Placing the mouse on the small label causes it start blinking:
> 
> ![Image](https://private-user-images.githubusercontent.com/841044/533979744-d6faaf1f-262f-471e-a422-15c495038907.gif?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjgxOTA5MzcsIm5iZiI6MTc2ODE5MDYzNywicGF0aCI6Ii84NDEwNDQvNTMzOTc5NzQ0LWQ2ZmFhZjFmLTI2MmYtNDcxZS1hNDIyLTE1YzQ5NTAzODkwNy5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMTEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDExMlQwNDAzNTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04ZGFhODQyODY1MDcxYjFjMzZiODcxNDQ0ZmU5ZWIzY2Y5NzNiYjMwODNhYmY4ZjBhYWM5YTA3ZTM5MTQwOWVlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.7QOgbDd-O5JQOEGIFEb0PX3DQc-VW-TAdb6eAybRsxo) [ ![Image](https://private-user-images.githubusercontent.com/841044/533979744-d6faaf1f-262f-471e-a422-15c495038907.gif?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjgxOTA5MzcsIm5iZiI6MTc2ODE5MDYzNywicGF0aCI6Ii84NDEwNDQvNTMzOTc5NzQ0LWQ2ZmFhZjFmLTI2MmYtNDcxZS1hNDIyLTE1YzQ5NTAzODkwNy5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMTEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDExMlQwNDAzNTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04ZGFhODQyODY1MDcxYjFjMzZiODcxNDQ0ZmU5ZWIzY2Y5NzNiYjMwODNhYmY4ZjBhYWM5YTA3ZTM5MTQwOWVlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.7QOgbDd-O5JQOEGIFEb0PX3DQc-VW-TAdb6eAybRsxo) ](https://private-user-images.githubusercontent.com/841044/533979744-d6faaf1f-262f-471e-a422-15c495038907.gif?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjgxOTA5MzcsIm5iZiI6MTc2ODE5MDYzNywicGF0aCI6Ii84NDEwNDQvNTMzOTc5NzQ0LWQ2ZmFhZjFmLTI2MmYtNDcxZS1hNDIyLTE1YzQ5NTAzODkwNy5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMTEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDExMlQwNDAzNTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04ZGFhODQyODY1MDcxYjFjMzZiODcxNDQ0ZmU5ZWIzY2Y5NzNiYjMwODNhYmY4ZjBhYWM5YTA3ZTM5MTQwOWVlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.7QOgbDd-O5JQOEGIFEb0PX3DQc-VW-TAdb6eAybRsxo) [ ](https://private-user-images.githubusercontent.com/841044/533979744-d6faaf1f-262f-471e-a422-15c495038907.gif?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjgxOTA5MzcsIm5iZiI6MTc2ODE5MDYzNywicGF0aCI6Ii84NDEwNDQvNTMzOTc5NzQ0LWQ2ZmFhZjFmLTI2MmYtNDcxZS1hNDIyLTE1YzQ5NTAzODkwNy5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMTEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDExMlQwNDAzNTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04ZGFhODQyODY1MDcxYjFjMzZiODcxNDQ0ZmU5ZWIzY2Y5NzNiYjMwODNhYmY4ZjBhYWM5YTA3ZTM5MTQwOWVlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.7QOgbDd-O5JQOEGIFEb0PX3DQc-VW-TAdb6eAybRsxo)
> 
> When the mouse leaves, there's still overlap:
> 
> ![Image](https://private-user-images.githubusercontent.com/841044/533979727-08d41e56-f79c-4ce2-9a4c-1d0ee2e46272.gif?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjgxOTA5MzcsIm5iZiI6MTc2ODE5MDYzNywicGF0aCI6Ii84NDEwNDQvNTMzOTc5NzI3LTA4ZDQxZTU2LWY3OWMtNGNlMi05YTRjLTFkMGVlMmU0NjI3Mi5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMTEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDExMlQwNDAzNTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zNTI3ZWJjOWZmM2IzYzI4YzE2ODZkYTI5OGRlZGQxYTE1ZWYyNTFjNGJhZTMzYjdlMGFjYjlhNDU1NTkyM2MyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.O4F08QGWel50gJmmmmTZPvHWTCIC8HCKpe4BVov5Ta4) [ ![Image](https://private-user-images.githubusercontent.com/841044/533979727-08d41e56-f79c-4ce2-9a4c-1d0ee2e46272.gif?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjgxOTA5MzcsIm5iZiI6MTc2ODE5MDYzNywicGF0aCI6Ii84NDEwNDQvNTMzOTc5NzI3LTA4ZDQxZTU2LWY3OWMtNGNlMi05YTRjLTFkMGVlMmU0NjI3Mi5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMTEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDExMlQwNDAzNTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zNTI3ZWJjOWZmM2IzYzI4YzE2ODZkYTI5OGRlZGQxYTE1ZWYyNTFjNGJhZTMzYjdlMGFjYjlhNDU1NTkyM2MyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.O4F08QGWel50gJmmmmTZPvHWTCIC8HCKpe4BVov5Ta4) ](https://private-user-images.githubusercontent.com/841044/533979727-08d41e56-f79c-4ce2-9a4c-1d0ee2e46272.gif?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjgxOTA5MzcsIm5iZiI6MTc2ODE5MDYzNywicGF0aCI6Ii84NDEwNDQvNTMzOTc5NzI3LTA4ZDQxZTU2LWY3OWMtNGNlMi05YTRjLTFkMGVlMmU0NjI3Mi5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMTEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDExMlQwNDAzNTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zNTI3ZWJjOWZmM2IzYzI4YzE2ODZkYTI5OGRlZGQxYTE1ZWYyNTFjNGJhZTMzYjdlMGFjYjlhNDU1NTkyM2MyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.O4F08QGWel50gJmmmmTZPvHWTCIC8HCKpe4BVov5Ta4) [ ](https://private-user-images.githubusercontent.com/841044/533979727-08d41e56-f79c-4ce2-9a4c-1d0ee2e46272.gif?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjgxOTA5MzcsIm5iZiI6MTc2ODE5MDYzNywicGF0aCI6Ii84NDEwNDQvNTMzOTc5NzI3LTA4ZDQxZTU2LWY3OWMtNGNlMi05YTRjLTFkMGVlMmU0NjI3Mi5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMTEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDExMlQwNDAzNTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zNTI3ZWJjOWZmM2IzYzI4YzE2ODZkYTI5OGRlZGQxYTE1ZWYyNTFjNGJhZTMzYjdlMGFjYjlhNDU1NTkyM2MyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.O4F08QGWel50gJmmmmTZPvHWTCIC8HCKpe4BVov5Ta4)
> 
> Please perform rigorous manual testing of your changes to ensure they're not adding quirks before asking for further reviews.
> 
> If we get to something that works, it would be nice to add at least 1 browser test for this.
> 
> Can you reply to the questions listed in the issue descirption and how you intend to solve each?
> 
> > I don't have a clear directive now, but I think the whole situation is a mess, here's the problems I see, I make suggestions at each point but I am open for better ideas:
> > 
> > * I am not sure where the hover overlay is coming from, whether is enabled by default or not
> > * It seems to me there's not an explicit ways to disable labels on the geographic map (they appears automatically after a certain level of zoom)
> > * If both are enabled, the overlap covers the label and they're quite redundant, ideally the labels should disappear if there's another overlay showing up on hover and by default only one of the two should be enabled (probably the label as it's not invasive)
> > * The current config options for controlling this behavior are: `showLabelsAtZoomLevel` for geo map and `showGraphLabelsAtZoom`, which is not consistent, I think we should rename `showLabelsAtZoomLevel` to `showMapLabelsAtZoom` (in the code we can continue supporting `showLabelsAtZoomLevel` for backward compatibility) and allow this to be `false` to disable lables altogether

1. Hover overlay origin
The overlay comes from the default ECharts tooltip. In this PR it’s made explicit and more predictable (confined to viewport, no hide delay).

2. Disabling map labels
Now they can be disabled completely if showMapLabelsAtZoom is set to false.
showLabelsAtZoomLevel is still supported internally for backward compatibility.

3. Overlap between labels and hover overlays
The main cause was labels being interactive. Both graph and map labels are now marked as silent, so tooltips always have interaction priority and overlap no longer causes UX issues.

4. Default behavior and redundancy
Labels remain the default, non-invasive layer (shown only above a zoom threshold), while tooltips remain the primary interactive overlay.

5. Config naming consistency
showLabelsAtZoomLevel (map) is effectively renamed to showMapLabelsAtZoom, aligning it with showGraphLabelsAtZoom, while keeping backward compatibility.

Dhruv-ub commented:
https://github.com/user-attachments/assets/765cb089-1d29-47b8-aca9-98b626b36794

Here is a demo of the fix in action. The labels now handle zooming correctly and do not overlap with tooltips.

Dhruv-ub committed:
[fix] Resolve redundant node labels and overlays

- Fixed hover overlay origin to be confined to viewport.
- Added logic to disable map labels completely if configured.
- Resolved overlap between labels and hover overlays (labels hide on hover).
- Ensured labels remain non-invasive while tooltips take priority.
- Renamed configuration options for consistency.
- Added regression tests for label visibility and hover interaction.

nemesifier reviewed (commented):
Can you add a browser test for this behavior?

nemesifier committed:
Merge branch 'master' into issues/454-fix-redundant-labels

nemesifier reviewed (changes requested):
Good progress @Dhruv-ub I like the result!  👍 

Here's a few problems found during my review and testing:

- I tried setting `showMapLabelsAtZoom` to false in one of the examples but I don't see any difference
- the README needs to be updated
- it doesn't look like we are maintaining backward compatibility with the old config option
- 1 test in the CI build is failing

I think with the issues above resolved we can merge.

Dhruv-ub committed:
[fix] Address review feedback:
- Fixed logic so 'showMapLabelsAtZoom: false' correctly disables labels.
- Restored backward compatibility for 'showLabelsAtZoomLevel'.
- Updated README documentation.
- Fixed CI test failure by updating mocks with 'fastDeepCopy'.
- Added regression test for disabled labels.

Dhruv-ub committed:
[fix] Refactor label visibility logic and resolve review comments

- Moved backward compatibility logic (showLabelsAtZoomLevel -> showMapLabelsAtZoom) to NetJSONGraph constructor in core.js.
- Refactored generateMapOption in render.js to remove redundant variables and use cleaner falsy checks.
- Updated mouseover/mouseout handlers to evaluate zoom dynamically.
- Removed redundant default value assignment in the render loop.
- Updated render tests to reflect logic changes.

nemesifier reviewed (changes requested):
I took a closer look and I have a number of doubts. See below.

Dhruv-ub commented on src/js/netjsongraph.render.js:
```diff
@@ -620,19 +654,56 @@ class NetJSONGraphRender {
+    // When a user hovers over a node, we hide the static label so the Tooltip
+    self.echarts.on("mouseover", () => {
+      if (!labelsDisabled && showLabel) {
+        self.echarts.setOption({
```
Yes, I also noticed this regarding the dynamic zoom check—I'll update the handler to evaluate that live.

nemesifier commented on src/js/netjsongraph.render.js:
```diff
@@ -179,6 +180,9 @@ class NetJSONGraphRender {
     const baseGraphSeries = {...configs.graphConfig.series};
     const baseGraphLabel = {...(baseGraphSeries.label || {})};
 
+    // Added this for label hover issue
```
```suggestion
    // Prevent redundant overlapping labels
```

nemesifier commented on .nvmrc:
```diff
@@ -1 +1 @@
-lts/krypton
+lts/krypton
```
ensure files always end with a blank line please

nemesifier reviewed (changes requested):
It's looking better, I may put my hands on it to improve a few things if needed, in the meantime can you please:

- add a selenium browser test which replicates the redundant node labels bug without the patch
- follow up with @coderabbitai until it approves the PR

See also my comments below.

Dhruv-ub committed:
Fix: Resolve redundant map labels on zoom and add regression test

Dhruv-ub committed:
[Test] Fix regression test variable name map vs graph

Dhruv-ub committed:
Merge branch 'master' into issues/454-fix-redundant-labels

Dhruv-ub committed:
[Fix] Correctly assign backward compatibility config to graph instance

Adityashandilya555 commented:
https://github.com/user-attachments/assets/43f395a1-1fc6-4b6e-8b56-4b1d35dccc9f

<img width="899" height="454" alt="Screenshot 2026-02-10 at 9 31 15 AM" src="https://github.com/user-attachments/assets/16cb5384-23c7-465d-aea9-c0223fef66c6" />
Almost everything is addressed. The only outstanding item is the .nvmrc missing trailing newline.

Dhruv-ub committed:
Addressed the line issue

Dhruv-ub committed:
[Fix] Addressed .nvmrc file issue

nemesifier commented:
@coderabbitai is this ready to merge?

Dhruv-ub commented:
> @coderabbitai is this ready to merge?

Will update as per its requirements and push asap

nemesifier committed:
Merge branch 'master' into issues/454-fix-redundant-labels

nemesifier commented on src/js/netjsongraph.render.js:
```diff
@@ -320,15 +324,19 @@ class NetJSONGraphRender {
         coordinateSystem: "leaflet",
         data: nodesData,
-        label: configs.mapOptions.nodeConfig.label,
+        animationDuration: 1000,
```
Why  did you add this?

I verified that this seems to affect only clusters, I think we should turn this off:

```suggestion
        // this affects clusters, we don't need the animation
        animationDuration: 0,
```

nemesifier commented on src/js/netjsongraph.render.js:
```diff
@@ -613,25 +626,65 @@ class NetJSONGraphRender {
+        });
+      }
+    });
+
```
```suggestion
```

nemesifier reviewed (changes requested):
I am sorry it took long for me to review this, it's been a tough period.

It looks mostly good, but on a closer look I still found some issues, see below.

I would appreciate a screen recording that shows how the UI issues have been addressed.

### 1. Performance: `setOption` called on every mouseover/mouseout (High)

**File:** `src/js/netjsongraph.render.js:629-667`

The `mouseover` and `mouseout` handlers each call `self.echarts.setOption()`
on every single hover event. `setOption` is not a cheap operation -- it triggers
ECharts' internal diff/merge and re-render pipeline. On a map with hundreds of
nodes, rapidly moving the cursor across multiple nodes will trigger this
repeatedly, potentially causing jank or frame drops.

**Suggestion:** Debounce the handlers, or track a boolean state flag and skip
the `setOption` call if the label visibility already matches the desired state.

```js
let labelsCurrentlyHidden = false;

self.echarts.on("mouseover", () => {
  if (!labelsCurrentlyHidden && showMapLabelsAtZoom && ...) {
    labelsCurrentlyHidden = true;
    self.echarts.setOption(/* hide */);
  }
});

self.echarts.on("mouseout", () => {
  if (labelsCurrentlyHidden && showMapLabelsAtZoom && ...) {
    labelsCurrentlyHidden = false;
    self.echarts.setOption(/* show */);
  }
});
```

### 2. Event handler accumulation on re-render (Medium)

**File:** `src/js/netjsongraph.render.js:629, 649`

`self.echarts.on("mouseover", ...)` and `self.echarts.on("mouseout", ...)` are
registered every time `mapRender` is called. If the user toggles between map
and graph mode via `switchMode` (see `netjsongraph.js:150-176`), `mapRender`
is called again, and new handlers are stacked on top of old ones. Over several
toggles this leads to duplicate handler invocations.

**Suggestion:** Call `self.echarts.off("mouseover")` /
`self.echarts.off("mouseout")` before registering, or use named functions and
`off` them first.

### 3. All labels hide on any node hover (Low/UX)

https://github.com/user-attachments/assets/46c03054-845d-4084-a182-dd62170c0755

**File:** `src/js/netjsongraph.render.js:629-667`

When hovering *any* node, *all* labels on the map disappear. The issue
description says "the labels should disappear if there's another overlay
showing up on hover." The current implementation is broader than necessary --
it hides every label rather than just the one overlapping with the tooltip.

It looks quite weird on a busy map. Can we only hide the specific node label  being hovered? That's what I was expecting, I wasn't expecting that all labels were hidden.

While testing for this issue, I found another issue: if you change the zoom level, labels are hidden until mouseover, see:

https://github.com/user-attachments/assets/57c5c019-87f3-456e-b735-f6ff439f6950

### 4. Config value captured at render time won't reflect dynamic changes (Low)

**File:** `src/js/netjsongraph.render.js:606`

```js
const {showMapLabelsAtZoom} = self.config;
```

This destructures the value once at render time. The `mouseover`, `mouseout`,
and `zoomend` handlers all close over this captured value. If a consumer
changes `self.config.showMapLabelsAtZoom` after initial render (e.g., via a
settings UI), the handlers won't see the updated value.

**Suggestion:** Read from `self.config.showMapLabelsAtZoom` directly in the
handlers instead of destructuring.

### 5. `animationDuration: 1000` added without clear reason (Low)

**File:** `src/js/netjsongraph.render.js:334`

The scatter series now has `animationDuration: 1000` added. This wasn't
mentioned in the issue or any commit message. It adds a 1-second animation to
node rendering which could impact perceived performance on large datasets.

### 6. `silent: true` also set on graph mode labels (Low)

**File:** `src/js/netjsongraph.render.js:178`

```js
baseGraphLabel.silent = true;
```

The issue is specifically about map mode label/tooltip overlap. Setting
`silent: true` on graph-mode labels means graph labels also won't respond to
mouse events. This may or may not be desired -- the issue description didn't
mention graph mode labels having overlap problems. If graph labels should remain interactive, this line should be removed.

**Question:** can you explain why this is added and what effect it has?

Dhruv-ub committed:
[fix] Map label visibility and render performance

- Optimize ECharts setOption to prevent hover frame drops
- Clean up event handlers on re-render to avoid duplication
- Hide only hovered node label and fix zoom visibility bug
- Read config state dynamically
- Disable cluster animation and remove graph label silent flag

Dhruv-ub commented:
https://github.com/user-attachments/assets/ad7f52da-732e-46fb-be83-05392a6b7d8c

Here is the testing video for the issue

nemesifier committed:
Merge branch 'master' into issues/454-fix-redundant-labels

nemesifier committed:
[chores] Removed dead code

nemesifier committed:
[fix] Allow showMapLabelsAtZoom: 0

nemesifier committed:
[fix] Show labels on first render

nemesifier committed:
[fix] Fixed label visibility when clustering is enabled

nemesifier reviewed (dismissed):
There were some bugs, I fixed any issue I encountered and it should be ready now.
@Dhruv-ub please double check when you can.
@coderabbitai review one last time, focus on real issues, approve if ready.

nemesifier committed:
[fix] Updated forgotten old config in netjsongraph.util.js

nemesifier reviewed (approved):
@coderabbitai review

nemesifier commented:
@coderabbitai review
